xdsl/frontend/code_generation.py

```diff
@@ -71,6 +72,105 @@ def visit_Assign(self, node: ast.Assign) -> None:
+        op_name: str = node.op.__class__.__name__
+
+        # Table with currently unsupported Python AST operators.
+        unsupported_python_AST_operator = {
+            "BitOr", "BitXor", "Div", "FloorDiv", "Mod", "MatMult", "Pow"
+        }
+
+        # Table with mappings of Python AST operator to Python methods.
+        python_AST_operator_to_python_overload = {
+            "Add": "__add__",
+            "BitAnd": "__and__",
+            "LShift": "__lshift__",
+            "Mult": "__mul__",
+            "RShift": "__rshift__",
+            "Sub": "__sub__",
+        }
+        if op_name in unsupported_python_AST_operator:
+            raise CodeGenerationException(
+                node.lineno, node.col_offset,
+                f"Unsupported binary operation {op_name}.")
+
+        self.visit(node.right)
+        rhs = self.inserter.get_operand()
+        self.visit(node.left)
+        lhs = self.inserter.get_operand()
+        if lhs.typ != rhs.typ:
+            raise CodeGenerationException(
+                node.lineno, node.col_offset,
+                f"Expected the same types for binary operation '{op_name}', "
+                f"but got {lhs.typ} and {rhs.typ}.")
+
+        # Look-up what is the frontend type we deal with to resolve the binary
+        # operation.
+        frontend_type = self.type_converter.xdsl_to_frontend_type_map[
+            lhs.typ.__class__]
+
+        op = OpResolver.resolve_op_overload(
+            python_AST_operator_to_python_overload[op_name],
+            frontend_type)(lhs, rhs)
+        self.inserter.insert_op(op)
+
+    def visit_Compare(self, node: ast.Compare):
+        # Allow a single comparison only.
+        if len(node.comparators) != 1 or len(node.ops) != 1:
+            raise CodeGenerationException(
+                node.lineno, node.col_offset,
+                "Expected a single comparator, but found "
+                f"{len(node.comparators)}.")
+        comp = node.comparators[0]
+        op_name: str = node.ops[0].__class__.__name__
+
+        # Table with currently unsupported Python AST cmpops.
+        unsupported_python_AST_cmpop = {"In", "Is", "IsNot", "NotIn"}
+
+        # Table with mappings of Python AST cmpop to Python methods.
+        python_AST_cmpop_to_python_overload = {
+            "Eq": "__eq__",
+            "Gt": "__gt__",
+            "GtE": "__ge__",
+            "Lt": "__lt__",
+            "LtE": "__le__",
+            "NotEq": "__ne__",
+        }
+
+        # Table with mappings of Python AST cmpop to xDSL mnemonics.
+        python_AST_cmpop_to_mnemonic = {
+            "Eq": "eq",
+            "Gt": "sgt",
+            "GtE": "sge",
+            "Lt": "slt",
+            "LtE": "sle",
+            "NotEq": "ne",
+        }
+
+        if op_name in unsupported_python_AST_cmpop:
+            raise CodeGenerationException(
+                node.lineno, node.col_offset,
+                f"Unsupported comparison operation '{op_name}'.")
+
```

Comment:
Same idea as above, in order to support different FrontendTypes supporting different sets, the failure should only happen when resolve_op_overload doesn't find the appropriate function in the FrontendType. Sorry for the back-and-forth!